akka-stream/src/main/scala/akka/stream/scaladsl/KillSwitch.scala

```diff
@@ -0,0 +1,119 @@
+ * ignored.
+ *
+ * After calling [[KillSwitch#abort()]] all materialized, running instances of all [[Flow]]s provided by the
+ * [[KillSwitch]] will fail their downstreams with the provided exception and cancel their upstreams
+ * (unless it finished or failed already in which case the command is ignored). Subsequent invocations of
+ * [[KillSwitch#shutdown()]] and [[KillSwitch#abort()]] will be ignored.
+ *
+ * The [[Flow]]s provided by the [[KillSwitch]] do not modify the passed through elements in any way or affect
+ * backpressure in the stream. All provided [[Flow]]s provide the parent [[KillSwitch]] as materialized value.
+ *
+ * This class is thread-safe, the instance can be passed safely among threads and its methods may be invoked concurrently.
+ *
+ * @param name A name that identifies the KillSwitch. Used for printing debug information. Two instances may have the
+ *             same name but this is not recommended practice.
+ */
+final class KillSwitch private (val name: String) {
+  private[this] val shutdownPromise = Promise[Done]
+  private[this] val _flow: Flow[Any, Any, KillSwitch] = Flow.fromGraph(new KillSwitchFlow)
+
+  /**
+   * After calling [[KillSwitch#shutdown()]] all materialized, running instances of all [[Flow]]s provided by the
+   * [[KillSwitch]] will complete their downstreams and cancel their upstreams (unless if finished or failed already in which
+   * case the command is ignored). Subsequent invocations of [[KillSwitch#shutdown()]] and [[KillSwitch#abort()]] will be
+   * ignored.
+   */
+  def shutdown(): Unit = shutdownPromise.trySuccess(Done)
+
+  /**
+   * After calling [[KillSwitch#abort()]] all materialized, running instances of all [[Flow]]s provided by the
+   * [[KillSwitch]] will fail their downstreams with the provided exception and cancel their upstreams
+   * (unless it finished or failed already in which case the command is ignored). Subsequent invocations of
+   * [[KillSwitch#shutdown()]] and [[KillSwitch#abort()]] will be ignored.
+   *
+   * These provided [[Flow]]s materialize to their owning switch. This might make certain integrations simpler than
+   * passing around the switch instance itself.
+   *
+   * @param reason The exception to be used for failing the linked [[Flow]]s
+   */
+  def abort(reason: Throwable): Unit = shutdownPromise.tryFailure(reason)
+
+  /**
+   * Retrurns a typed Flow of a requested type that will be linked to this [[KillSwitch]] instance. By invoking
+   * [[KillSwitch#shutdown()]] or [[KillSwitch#abort()]] all running instances of all provided [[Flow]]s by this
+   * switch will be stopped normally or failed.
+   *
+   * @tparam T Type of the elements the Flow will forward
+   * @return   A reusable [[Flow]] that is linked with the switch. The materialized value provided is this switch itself.
+   */
+  def flow[T]: Flow[T, T, KillSwitch] = _flow.asInstanceOf[Flow[T, T, KillSwitch]]
+
```

Comment:
@viktorklang In general I like the idea of reducing concepts, but unless I have misunderstood things your proposal will not support one "global" switch for all materialized instances, since the promise is returned as materialized value instead of being provided by the user.